### PR TITLE
add fetchTasks() and executeTask() to tasks API

### LINKED ISSUE #5342: [plug-in][tasks] Add ability to fetch all tasks
### Description
Add [tasks.fetchTasks()](https://github.com/microsoft/vscode/blob/5a4eab812bd485942061f71af39f3eabc52a1035/src/vs/vscode.d.ts#L5059) API to provide ability to get all tasks available in the systems

### PR DESCRIPTION
**Pull request by elaihau**
- This pull request adds the support of executing a vscode.Task, and fetching tasks by type and schema version to the plugins API.
- resolved #5342

Signed-off-by: Liang Huang <liang.huang@ericsson.com>

#### How to test

This pull request can be tested with adding a vscode extension to Theia:

```ts
import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {

	console.log('Congratulations, your extension "fetchtest" is now active!');

	let disposable = vscode.commands.registerCommand('extension.helloWorld', () => {
		// The code you place here will be executed every time your command is executed

		// Display a message box to the user
		vscode.window.showInformationMessage('Test Fetch Task and Execute Task!');

		// vscode.tasks.fetchTasks({ type: 'shell' }).then((value) => {
		// 	console.log('shell fulfilled', value);
		// }, (reason) => {
		// 	console.log('rejected', reason);
		// });

		vscode.tasks.fetchTasks({ type: 'npm' }).then((fetchedTasks) => {
			console.log('npm tasks found: ', fetchedTasks.length);
			for (const task of fetchedTasks) {
				console.log('\n\nlooking at ', task.name, '\n\n');
				vscode.tasks.executeTask(task);
			}
		}, (reason) => {
			console.log('fetch task was rejected', reason);
		});

		// const myTask = new vscode.Task(
		// 	{ type: 'shell' }, // task definition
		// 	vscode.TaskScope.Workspace,
		// 	'test',
		// 	'testExtension',
		// 	new vscode.ShellExecution('ls -alR', { cwd: '/home/elaihau/dev' })
		// );
		// vscode.tasks.executeTask(myTask);
	});

	context.subscriptions.push(disposable);
}

export function deactivate() { }
```

- the extension can be locally packaged by `vsce` https://code.visualstudio.com/api/working-with-extensions/publishing-extension
- the extension, when added as a plugin, should add a command `Test Fetch Task and Execute Task` to Theia
- by running the command contributed by the extension, we are able to test if the tasks can be fetched by the new API, also, if the fetched tasks can be run.

**For reviewer's convenience I uploaded the source code** of my extension to https://github.com/elaihau/fetchtest

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

### CODE REVIEW

File: packages/plugin/src/theia.d.ts [hunk 1]
```diff
@@ -8202,6 +8202,19 @@ declare module '@theia/plugin' {
...
 
+    export interface TaskFilter {
+		/**
+		 * The task version as used in the tasks.json file.
+		 * The string support the package.json semver notation.
+		 */
+        version?: string;
+
+        /**
+     * The task type to return;
```
[the remark this answers is not in the document]
Comment (elaihau) [reply packages/plugin/src/theia.d.ts]: i changed it to `the type of tasks to return.`

File: packages/task/src/browser/task-service.ts [hunk 2]
```diff
@@ -303,7 +303,7 @@ export class TaskService implements TaskConfigurationClient {
      * Runs a task, by the source and label of the task configuration.
      * It looks for configured and detected tasks.
      */
-    async run(source: string, taskLabel: string): Promise<void> {
+    async run(source: string, taskLabel: string): Promise<TaskInfo | undefined> {
```
[the remark this answers is not in the document]
Comment (elaihau) [reply packages/task/src/browser/task-service.ts]: correct me if i am wrong:
before this change, the client does not expect this function to return anything.
so, after this change, the return value from this function wouldn't break any clients.

therefore i don't think it is `breaking`. It is a breaking change if I change the return type from `Promise<TaskInfo>` to `Promise<void>`. What do you think?

File: packages/plugin-ext/src/plugin/tasks/tasks.ts [hunk 3]
```diff
@@ -109,6 +109,27 @@ export class TasksExtImpl implements TasksExt {
         return this.createDisposable(callId);
     }
 
+    async fetchTasks(filter?: theia.TaskFilter): Promise<theia.Task[]> {
+        const taskVersion = filter ? filter.version : undefined;
+        const taskType = filter ? filter.type : undefined;
+        const taskDtos = await this.proxy.$fetchTasks(taskVersion, taskType);
+        return Promise.resolve(taskDtos.map(dto => converter.toTask(dto)));
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/tasks/tasks.ts]: just fyi: you don't need `Promise.resolve` or `Promise.reject` in async context

just `return` or `throw`

Comment (elaihau) [reply packages/plugin-ext/src/plugin/tasks/tasks.ts]: thank you for the review. fixed.

File: packages/task/src/browser/task-service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vince-fugnitto) [reply packages/task/src/browser/task-service.ts]: Yes I agree, it shouldn't be breaking :)

**Comment by RomanNikitenko** [comment]
@elaihau @akosyakov 
I'm on vacation and going to be offline next 2 weeks.
thank you for understanding!

**Comment by elaihau** [comment]
@RomanNikitenko thank you for the heads up ! Have fun :) 

@vince-fugnitto maybe i can walk you through the code and explain how to test it this Friday (Sep 6) ?

File: packages/plugin/src/theia.d.ts [hunk 4]
```diff
@@ -8213,6 +8226,23 @@ declare module '@theia/plugin' {
...
         export function registerTaskProvider(type: string, provider: TaskProvider): Disposable;
 
+        /**
+         * Fetches all tasks available in the systems. This includes tasks
+         * from `tasks.json` files as well as tasks from task providers
+         * contributed through extensions and plugins.
+         *
+         * @param filter a filter to filter the return tasks.
+         */
+        export function fetchTasks(filter?: TaskFilter): Thenable<Task[]>;
```
Comment (mmorhun) [inline packages/plugin/src/theia.d.ts]: Please use `PromiseLike` insted of `Thenable`

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto maybe i can walk you through the code and explain how to test it this Friday (Sep 6) ?

Yes, that sounds fine with me :)

**Comment by mmorhun** [comment]
I've tested the changes for the following cases:
- Fetch tasks:
  * Fetch tasks from `tasks.json` inside `.theia` folder of specific type
  * Fetch tasks of specific type which are provided by a `TaskProvider` contributed by plugin.
  * Mix of two previous cases: fetch tasks of specific type when they are provided by `tasks.json` and some by a `TaskProvider` from within a plugin.
- Run tasks:
  * Fetch tasks of specific type, found the task with specifi label, change its configuration and run altered task.

All described above cases worked for me.

Resource used for testing of this PR:
<details>
<summary>Source code of the test plugin </summary>

```typescript
import * as theia from '@theia/plugin';

export function start(context: theia.PluginContext) {
    const listTAsksTestCommand = {
        id: 'list-tasks-test-command',
        label: "Test list tasks of type"
    };
    context.subscriptions.push(theia.commands.registerCommand(listTAsksTestCommand, (...args: any[]) => {
        theia.window.showInformationMessage('On list tasks');

        theia.tasks.fetchTasks({ type: 'ctask' }).then((tasks: theia.Task[]) => {
            console.log('IN CALLBACK:', tasks);
            for (const task of tasks) {
                console.log('TASK:', task.name);
            }
        });
    }));

    const runTaskTestCommand = {
        id: 'run-task-test-command',
        label: "Test alter task"
    };
    context.subscriptions.push(theia.commands.registerCommand(runTaskTestCommand, (...args: any[]) => {
        theia.window.showInformationMessage('On run altered task');

        theia.tasks.fetchTasks({ type: 'ctask' }).then((tasks: theia.Task[]) => {
            console.log('IN RUN TASK:', tasks);
            for (const task of tasks) {
                if (task.name.startsWith('Echo hello world')) {
                    if (!task.execution) {
                        task.execution = {};
                    }
                    if (!task.execution.options) {
                        task.execution.options = {};
                    }

                    (task.execution as theia.ShellExecution).commandLine = 'sleep 1 && ls -la';
                    task.execution.options.cwd = "/tmp";

                    theia.tasks.executeTask(task);

                    return;
                }
            }
        });
    }));

    theia.tasks.registerTaskProvider('ctask', new CTaskProvider());
}

class CTaskProvider {
    async provideTasks(): Promise<theia.Task[]> {
        const task1 = new theia.Task({ type: 'ctask' }, theia.TaskScope.Global, 'first task', 'the-extension');
        const task2 = new theia.Task({ type: 'ctask' }, theia.TaskScope.Workspace, 'second task', 'ctask');
        return [task1, task2];
    }

    async resolveTask(task: theia.Task): Promise<theia.Task> {
        return task;
    }
}

export function stop() { }

```

</details>

<details>
<summary>tasks.json</summary>

```json
{
    "tasks": [
        {
            "type": "abcd",
            "label": "rnodejs-hello-world:run",
            "command": "yarn",
            "target": {
                "workingDir": "/projects/theia"
            },
            "previewUrl": "${server.3000/tcp}"
        },
        {
            "type": "ctask",
            "label": "Echo hello world",
            "command": "echo \"hello world\"",
            "target": {
                "workingDir": "/projects",
                "containerName": "theia-dev"
            }
        }
    ]
}
```

</details>

**Comment by elaihau** [comment]
Thank you @mmorhun  !

**Comment by elaihau** [comment]
Thank you for the review @mmorhun !

I updated the code to address your review comments.

**Pull request merged**